transient/config.py

```diff
@@ -0,0 +1,261 @@
+
+    def __setitem__(self, key: Any, value: Any) -> None:
+        super(Config, self).__setitem__(key, value)
+        self.__dict__.update({key: value})
+
+    def __delattr__(self, item: Any) -> None:
+        self.__delitem__(item)
+
+    def __delitem__(self, key: Any) -> None:
+        super(Config, self).__delitem__(key)
+        del self.__dict__[key]
+
+
+class TransientConfigSchema(Schema):
+    """Defines the schema for Transient-run arguments and validates the fields
+       during deserialization
+    """
+
+    config = fields.Str(allow_none=True)
+    copy_in_before = fields.List(fields.Str(), allow_none=True)
+    copy_out_after = fields.List(fields.Str(), allow_none=True)
+    copy_timeout = fields.Int(allow_none=True)
+    image = fields.List(fields.Str(), allow_none=True)
+    image_backend = fields.Str(allow_none=True)
+    image_frontend = fields.Str(allow_none=True)
+    name = fields.Str(allow_none=True)
+    prepare_only = fields.Bool(allow_none=True)
+    qemu_args = fields.List(fields.Str(), allow_none=True)
+    qmp_timeout = fields.Int(allow_none=True)
+    shutdown_timeout = fields.Int(allow_none=True)
+    ssh_command = fields.Str(allow_none=True)
+    ssh_bin_name = fields.Str(allow_none=True)
+    ssh_port = fields.Int(allow_none=True)
+    ssh_timeout = fields.Int(allow_none=True)
+    ssh_user = fields.Str(allow_none=True)
+    ssh_console = fields.Bool(allow_none=True)
+    ssh_with_serial = fields.Bool(allow_none=True)
+    shared_folder = fields.List(fields.Str(), allow_none=True)
+
+    # marshmallow's decorator post_load() is untyped, forcing create_args()
+    # to be untyped. Therefore, we ignore it to silence the type checker
+    @post_load  # type: ignore
+    def create_config(self, data: Dict[Any, Any], **kwargs: Dict[Any, Any]) -> Config:
+        """Returns the Config dictionary after a schema is loaded and validated
+        """
+        return Config(**data)
+
+
+def replace_hyphens_with_underscores_in_dict_keys(
+    dictionary: Dict[Any, Any]
+) -> Dict[Any, Any]:
+    """Replaces hyphens in the dictionary keys with underscores
+
+       This is the expected key format for TransientConfigSchema
+    """
+    final_dict = {}
+    for k, v in dictionary.items():
+        # Perform this method recursively for sub-directories
+        if isinstance(dictionary[k], dict):
+            value = replace_hyphens_with_underscores_in_dict_keys(v)
+            final_dict[k.replace("-", "_")] = value
+        else:
+            final_dict[k.replace("-", "_")] = v
+
+    return final_dict
+
+
+def parse_config_file(config_file_path: str) -> MutableMapping[str, Any]:
+    """Parses the given config file and returns the contents as a dictionary
+    """
+    with open(config_file_path) as file:
+        config_file = file.read()
+
+    try:
+        parsed_config_file = toml.loads(config_file)
+    except RuntimeError as error:
+        print(error, file=sys.stderr)
+        sys.exit(1)
```

Comment:
We should not `sys.exit`, except within `cli.py`. If someone is trying to use transient as a library, this will kill their program.

Reply:
I also don't think we should suppress the exception. There is a general question of how to display errors if the user is coming from the CLI, but that should be handled by catching the exceptions there.

Reply:
Replaced sys.exit(1) in config.py with raise in functions:

    load_config_file()
    parse_config_file()
    create_transient_run_config_from_cli_args()

That way, these functions can still print useful error messages during exception handling and re-raise the exception.